### DISCUSSION TITLE
5150 Emulation flaws

### DISCUSSION THREAD

**Discussion by BigAlUK99** [86Box/86Box#1724] [General]
There's quite a few flaws in the IBM PC 5150 emulation - some of which should be trivial to sort, but some are subject to _really_ understanding the relationship between the 5-slot limit, motherboard switches, BIOS release dates and available ISA cards at the time these combinations existed.

---
1.  When selecting the IBM PC (1981),  the 02/24/81 BIOS gets combined with the C1.10 BASIC ROM.  This is **incorrect**.  The C1.10 BASIC ROM should _only_ be used with the 10/27/82 BIOS.  Whilst this may merely be down to incorrect content of the ROM's ZIP file, it's relationship to the 86Box settings menu shows that this aspect has not been properly understood or accounted for.

There are in fact ONLY FOUR valid combinations for the 5150:
-  02/24/81 BIOS + C1.00 BASIC ROM  : The very first PC's shipped in August 1981
-  10/19/81 BIOS + C1.00 BASIC ROM  : The next batch shipped from October 1981
-  10/27/82 BIOS + C1.?? BASIC ROM  : This is Rev.A - Fully populated 64K RAM on old board
-  10/27/82 BIOS + C1.10 BASIC ROM  : This is Rev.B - Min. of 64K RAM (of 256K) on new board

---
2. If the 10/27/82 BIOS is fitted, it is implied that the memory switches in SW1 are set at max (i.e. 64K).  This means the other switch bank is set for the remaining RAM, irrespective of whether that resides on the system board or in ISA slots.   On a Rev.A system you are unlikely to find a full 640K because of the limited number of ISA slots and the size of available ISA RAM cards.  On a Rev.B the first 256K RAM did not require ISA slots to be used, so reaching 640K is easier.  In 1983, IBM started releasing much larger ISA RAM cards, so people no-longer needed to rely upon 3rd-party cards.   However, _before_ Rev.A, the older 1981 BIOS ignored the top switch in the SW2-bank, so there could never be more than 512K RAM until the 1982 BIOS (i.e. Rev.A and Rev.B motherboards).

It appears that 86Box is not setting the switches correctly in regard to the above rules.  The validity of the switches depends heavily upon which BIOS is being used.  The memory-related options in the 86Box settings (for system and ISA based RAM) therefore make little sense and don't even appear to have the correct effect (if any) upon the switches.  Ergo the BIOS will not know the correct RAM size, and thus neither will DOS.

Note that the switches are handled slightly differently on an IBM 5160, so the algorithm for these switches is dependant upon the emulated machine type (and hence the BIOS).  Furthermore, unless this is all corrected, there will be bugs in the behaviour of all software that relies upon the BIOS to correctly identify the memory configuration of the 5150.  Be warned: to get this stuff right requires reading and fully understanding _all_ of the releases of the corresponding IBM documentation, not just "the latest".

---
3. There is likewise a good deal of confusion regarding the LPT port provision.  Again, it can also be related to the BIOS date used, but only in the sense of understanding what ISA cards were available when these BIOS dates were being used.

- The early 1981 BIOS implies that there were only two types of graphics cards available (CGA and MDA).  The MDA _always_ had its own LPT port and so only 1 ISA slot was needed to have graphics + LPT1.  With the CGA, however, you needed a separate LPT card (which only had _one_ LPT port).  To get THREE such ports - which 86DOS insists upon in an all-or-nothing manner - means that you would be using 4 out of the 5 slots (FDD, MDA, LPT2, LPT3) or all 5 slots (FDD, CGA, LPT1, LPT2, LPT3).   This would limit what options could be used for RAM above 64K or whether a serial card or game port card could be used.   The 3rd-party cards that combined functions were not really available when the JUNE 1981 BIOS was being used!

- If the AUGUST 1981 BIOS is fitted, then this implies there are more options.  It was fairly unlikely that the older BIOS would have still been fitted if someone was spending money to actually use these 3rd-party options - and even less likely if they were using IBM parts!  The new options included Plantronics video (also with its own LPT port), and certain game/serial/parallel/memory combination cards (not from IBM).  By 1982, you would have Hercules (also with its own LPT), and a raft of EGA cards (many without an LPT).  By 1983, IBM were also supplying multi-card options.   

- However, it must be understood that if either 1981 BIOS was still fitted, then this meant that these ISA-based devices could not have their own BIOS chips, which rules out EGA, VGA, and many other specialist cards, and, of course, hard drives.  This is because these BIOS chips need to be executed from the IBM BIOS - which didn't happen until 1982.

- Finally, in 1982/83, IBM came up with the 5161 expansion box.  And this could only be used with Rev.A and Rev.B systems, because the 1981 BIOS codes did not know how to drive it   So having a 1981 BIOS means you cannot have a 5161.  Of course, if you do have a 5161, the the slot-related limitations change, because a 5161 has 8 additional slots.  HOWEVER, one of the 5 original and one of the 8 new slots are taken up by connecting this box - so you are only adding an extra 6 useful slots.

These considerations show that the "slot" related settings in 86Box are somewhat naive at best.

---

**Comment by BigAlUK99** [comment]
But aside from all that - I quite like the way 86Box is going.  :)

**Comment by OBattler** [comment]
You have very good points there. I'll be looking into that after v2.00 is out.

The current design has no concept of ISA slots. The concept of slots is currently only used for MCA and PCI, and even that internally - the user does not see the slots. Eventually this should be changed to have ISA/MCA/VLB/PCI slot selection boxes and putting cards into them. This is likely going to require a new overhaul of the Settings dialog, but it would make add-on card emulation considerably more realistic (and for ISA, 8-bit and 16-bit slots should be distinguished, as there might even be machines that have both types).

As for the IBM BASIC - I'd need to find ROM's for the other versions of it, then I could fix that problem.

**Comment by BigAlUK99** [comment]
I have a ton of research and test material that may be of use when you come to it  - you may have noticed my posts on the PCem forum and the MAME bug report in this regard.  I use a program written in Cassette BASIC to verify whether the emulation is providing enough (and feasible) inputs to the BIOS.  It's likely to be very useful as a convenient test tool (though it still has a few behavioural flaws when the emulation proves to be unfeasible - I shall clean that up eventually - though I'm limited by the code having a design aim to also run in less than 16K of RAM when loaded via the cassette interface).  

So far, the only emulator I've found that ticks (almost) all the boxes in regard to (fairly) accurate emulation of the 5150 is Hampa's PCE - as long as the CFG file is adequately and sensibly configured.

The code of my BASIC test program reads (what should be) the motherboard switches, and the BIOS data areas set up when the BIOS runs, and (what should be) the ports of the graphics card.  As yet I haven't found an emulation that delivers 5161 signals, or anything that can directly identify slot usage.  Thus the program uses the data available, plus historical knowledge of the ISA cards available at a given time and the relationship of the BIOS date to all of this including the potential system board (PC1, RevA or RevB).   It then computes more or less what the emulation represents in terms of the 5150 model (quoting a derived model number that is historically adjusted) and what it deduces that the BIOS has detected in terms of hardware.  In this regard, one of the most intricate calculations made (re: feasibility of emulated machine) is that of how the system board and ISA-located RAM is provided, i.e. which cards could have been used and whether it's configuration could have feasibly existed at the time, or represents a post-delivery "upgrade" requiring "future" or "non-IBM" parts.  Generally speaking it is this final part (identifying less feasible configurations) that is still a bit flawed in my code (and there's so little room left in the 16K).

It has been a bit of a labour of love, primarily driven by the fact that the 5150 of 1981 to 1984 had such an enormous impact upon the direction of computing and is therefore a fascinating historical artefact.  It therefore in my view deserves the most accurate emulation possible, given the propensity for real-world capacitors to randomly explode.

_(copy and paste links)_
- https://1drv.ms/u/s!AnT0q0DRuXBNz0GJXoz0amhpIU0x  - sample picture of an IBM PC 5150-013 result.
- https://1drv.ms/u/s!AnT0q0DRuXBNz0QKfcBd4OtglLCA - sample of an IBM PC 5150-004 before the model numbers changed (decided upon the BIOS date)
- https://1drv.ms/u/s!AnT0q0DRuXBNz0XpS-yttpnU7maF - this one is interesting because it is PCem rather than PCE, and it demonstrates (a) detection of the Plantronics card, (b) the infeasibility of insisting upon all that I/O always being present (already mentioned this to Sarah), the way I have to handle the absence of 5161 signals by assuming two possible solutions (RevA or RevB systems) and of course, finally, the top line that says this is not a feasible emulated machine - which leads into the code I haven't finished yet because it is very hard to fit it into 16K of RAM :)

The MAME emulation is not worth bothering with yet, because they simply don't provide the system card switches in a coherent manner for RAM yet - although kudos to them for emulating the 8087 (and signalling it correctly at least) and for allowing emulation of a dual-graphics MDA+CGA system (although I'm not sure enough about its implementation yet to incorporate detection of this into the BASIC program).  They also upgraded their Cassette Port emulation (at my prompting), so that it is actually even better than the one in PCE (much more user friendly).

The current version of my BASIC test software (with aforementioned flaws) is here:
- https://1drv.ms/u/s!AnT0q0DRuXBNz0YDZhUQar1ggt5c - This is on a 160K Single Sided IMG floppy (won't be readable with WinImage or ImDisk, but most emulators will read it).  I chose this format because it is the only one readable by PC-DOS 1.0 (and later).  However, you should not really write to it if a later DOS is loaded.
- https://1drv.ms/u/s!AnT0q0DRuXBNz0eJ_JYidQdKOYVm - This is on a PCE-compatible CAS Tape emulation, assuming an emulator can read it, and for when you cannot use a floppy because less than 48K RAM is fitted (DOS has needs too!).  Be aware that this might not be the exact same code as on the floppy, since both represent work-in-progress.

**Comment by OBattler** [comment]
Well, on 86Box, serial and parallel ports can be disabled, if i do that, on the 1982 PC, it no longer says unfeasible. How is the 5161 signalling done?

**Comment by BigAlUK99** [comment]
If you look at the listing of the BASIC code, the last 8 or so lines hold two subroutines: The 5161 I/O test followed by the Plantronics Graphics card test.    The EM5161 global variable after returning is set to 0 if no 5161 signal was detected and to 1 if it was.  This is based on what I gleaned from the BIOS listing in the Technical References for the RevA and RevB systems.

If you are using the IBM Personal Computer Technical Reference manual of April 1983, then if you look on page A-8, you will see that the expansion box is in theory enabled by writing a 1 to port 213H.   Then, on page A-13, you'll see the actual test, which is to toggle a bit pattern to port 210H.  They use words of 5555H and AAAAH.    However, I have also since noticed that enabling the box may be insufficient - since on page A-91 you can see code that turns the box on (FFH to port 213) and off (0 to port 213H).  But that may be "too much".  I think - though not certain - that the test is done whilst merely "enabling" rather than "powering".  Still, it is quite hard to make an emulator go BANG!   :)
 
Incidentally, you can find these references by searching a PDF of this document for the phrase "EXPANSION BOX".   https://1drv.ms/b/s!AnT0q0DRuXBNz0kMFAsd8ib9yAVQ

If you are using the IBM Personal Computer Technical Reference manual of April 1984,  you will however see these items on pages 5-36 and 5-41, but not the part that is equivalent to A-91 in the other manual!  This is because A-91 in the other 1983 manual lists the hard disk's BIOS, which is missing in the 1984 manual (Gee, thanks IBM).   So that's why I guessed that writing a '1' rather than 'FF' would be sufficient to enable the test.  This 1984 Manual is here: https://1drv.ms/b/s!AnT0q0DRuXBNz0hcmVS8sHpDmgWK

Of course, if you look in the IBM Personal Computer Technical Reference manual of August 1981, you won't find _any_ of this.  Just like you will not find code that reads System Switch Bank SW2 switch 5 - which is why the pre-1982 BIOS's cannot have more RAM than 512K from an electrical viewpoint - ignoring for now that at that time you also wouldn't have had enough ISA slots either, because IBM were not making larger memory cards yet.  For your edification, this 1981 manual is here:  https://1drv.ms/b/s!AnT0q0DRuXBNz0rT7QG_i901PWGi

As a side-note, this is also why 86-DOS and PCDOS1.0 "blow up" when fitting too much memory - I think the limit they assume in the DOS code is actually lower still.  So if your emulator user gets this wrong by simply fitting "too much" memory, these early versions of DOS stop functioning before the BIOS does.  Not really a bug - just lower expectations that were rapidly exceeded!  These were the days of pragmatic programmers shaving off bytes from their code.  :)   Even 256K was impressive then - well, for a month or two at least.

On a final note, my BASIC code does not really do the test correctly!   Presently all it is really doing is confirming that none of the emulators are doing anything with these two ports 213H and 210H.   If this test were to be accurate, I should be using WORDS not BYTES - though I have no idea if BYTES would work in the real world.  Why am I doing this?  Because I'm up against my limit of 16K of RAM for my BASIC program (heheheh)!  Every statement I can reduce is precious!

So OBattler, there you have it :)

If you want, there is also a manual for the actual 5161 here: https://1drv.ms/b/s!AnT0q0DRuXBNz0udSS_k_cZy7bIg  - assuming you have the patience to emulate all the switches, bells and whistles in this box.  The really interesting stuff is on page 9 (a table of IO Ports)  and page 5 (extender card and switches).

But I shall leave that to you. :)

Incidentally, it is probably all these aspects that have led people to rely upon the XTIDE driver to fit emulated hard drives, instead of using the real NVR-writing code on IBM's hardware test disks.  That and the confusion potentially surrounding all of those pictures showing 5150 boxes with hard-drives (they're not IBM - they're 3rd-party - probably).  When in fact, if you had a 5150, IBM actually expected you to buy a 5161 with a 10MB Hard drive already fitted, and then optionally add a second 10MB HD also into the 5161 box if you could afford all that outlay £10K+ £20K+ ? for a stonking total of 20MB.  Let alone all that desktop space for these two boxes.  Stacking them was awkward, because the monitor would be a little high!  :)    But whatever - you will know far more than I do what it would take to remove the (unrealistic) XTIDE from the equation and use a real IBM HD BIOS (as is in the April 1983 Manual).

**Comment by BigAlUK99** [comment]
Incidentally,  I have mentioned on other forums (and no-one has given any clarification yet) that the date records in BIOS listings in the 1983 and 1984 manuals do not appear match each other - which implies at least one of these listings was not what actually went into the 1982 BIOS ROM.   Some post-event rationalization of a total lack of configuration control at IBM?  You decide!   Heaven forbid that there were actually _four versions_ released!

So anyways - it is worth bearing this in mind and confirming via some kind of debugging aid that the actual ROM-held BIOS code is doing what these listings imply  -  just sayin'   :)

**Comment by OBattler** [comment]
Well, for what it's worth, I haven't used XTIDE in a while, it's still there, but both PCem and 86Box have long since added the IBM Fixed Disk Controller (that PCem calls Xebec), which uses the IBM HD BIOS. So that has long been solved. :p

As for the 5161, I guess I need to emulate the RAM addresses (how much memory at Fxxx is assigned to the 5161?) and I/O ports. I did look at the 5161 reference manual, it shouldn't be that difficult.

But right now, I have decided to add the 8087 first, and am a bit stuck there - CHKCOP attempts to execute the FSTSW AX instruction which is 287+, and hangs if it's not present. I need to figure out what its opcode (DF E0) does on the 8087. With the instruction present, there's no hang but CHKCOP says it's fouind an "Intel 80???" coprocessor. I asked around to see if anyone has a ready binary of the latest PCE, so I can see if CHKCOP works there.

**Comment by BigAlUK99** [comment]
Brill! Although I should perhaps mention that the 8087 was a retro-fit to any 5150 before '82/83 - so it would be nice to reflect that historic info in your menu-ing system such that it is not a default and always optional :)

Unfortunately, I've never used a real physical 5161 either and only recently learned about it.  So my knowledge is completely retrospective and depends on the surviving paperwork.  Whilst a lot of what I relate here I vaguely remembered from the time; the details are all obtained from research over the last 6-months, so I'm probably in the same boat as you.

I didn't know about the Xebec stuff being available to play with already!  Is there something I can use as a tutorial for setting it up?   Will it work with HD-setup tools in the IBM Diagnostics floppies?   Is it correctly limited to 1982 BIOS usage in the emulator menus?  Indeed, am I even correct in believing that in a real-world scenario, the HD BIOS cannot even be physically invoked without the system BIOS having the code to do so? (Hence the 1982 BIOS prerequisite).  Just sayin': I'm aware that with an emulation you might be able to trick the execution of the HD BIOS rather than having the actual system BIOS search and invoke it.

**Comment by BigAlUK99** [comment]
By the way, I do not have access to a physical 5150 machine, so if you do, or you know someone who does, it would be really great to push a copy of the BASIC code to them and confirm it correctly identifies real-world 5150 hardware.   Preferably before those capacitors explode from old-age :)

**Comment by BigAlUK99** [comment]
I shall do some research on the 8087 question.  I vaguely remember having some very old ASM code somewhere that tests for things like the presence and nature of co-processor support (rather than just relying on the motherboard switch).  And I have several hardcopy books in the loft somewhere from those days, so maybe something there...  get back to you on this later.

**Comment by BigAlUK99** [comment]
So, the FNSTSW AX info I have is this:

First off, as you already say, it is actually an 80287 instruction - the 8087 device could not execute it, having only the circuitry to store directly into memory rather than injecting it into AX register of the main processor!   So that may be your problem right there?  See https://1drv.ms/u/s!AnT0q0DRuXBNz06Ze73F5AduZedb, the seventh line of text from the top.

What I am getting at, is that this may not be compatible with an 8088 processor, as exists in the 5150.  Especially given that the 8088 uses an 8-bit bus, not a 16-bit bus, so the 8088 will possibly need to know what to do here and may not be expecting this.  Which could explain your error message of "Intel 80???" coprocessor.  But then I know nothing about your test harness here and whether you are indeed executing it upon an 8088 8-bit bus system.

**Comment by BigAlUK99** [comment]
Some other potentially useful info from the same source (a long-ago downloaded web page from 16 years ago) are the discussions about the other "management" instructions under the general heading "Instructions related to the FPU internals" can be found on the web source (still available!)

See http://www.ray.masmcode.com/tutorial/
and specifically: http://www.ray.masmcode.com/tutorial/fpuchap3.htm

The page I linked to you as a JPG from my disk-archive is still there as: http://www.ray.masmcode.com/tutorial/fpuchap3.htm#fstsw

You probably need to sift through this and other sources to make sure you are emulating the correct behaviour regarding interaction between what was an 8088 processor coupled with what should really be an 8087 (not an 80287).   I shall have a further look to see what I can find regarding this.

**Comment by BigAlUK99** [comment]
As an aside...
This page was amusing:   https://webcache.googleusercontent.com/search?q=cache:f_Ehm5T2qqYJ:https://www.oreilly.com/library/view/pc-hardware-in/059600513X/ch04s02.html+

because it underlines the point that the FPU must match the CPU, and in its example has the following anecdote:

_"The 486SX was in fact a 486DX with the FPU disabled. Intel also sold the 487SX, which was actually a full-blown 486DX. Installing a **487SX** in the coprocessor socket simply disabled the existing 486SX."_  

Well, I was amused, anyway (simple minds).   Electronics design for the establishment of income by superfluous redundancy.  

Translating this into English works like this:
- We have sold you a broken chip
- You can always buy a working chip from us and fit it into the extra socket we have already charged you for, and you don't even have to remove the broken chip.
- We only did this because we knew we would most likely manufacture rather a lot of broken chips
- We can also sell you the same identical chip twice, but one of them really won't work, ...maybe ...you'll never know because we switch it off.

**Comment by BigAlUK99** [comment]
This page (http://www.cpu-world.com/CPUs/8087/index.html) explains neatly (i.e. simply) how the 8087 interacts with its potentially compatible CPU compatriots (8086, 8088, 80186 and 80188) by monitoring the address bus upon detecting an "escape" instruction byte, but haphazardly not actually synchronizing with the data bus - more likely to be a bigger issue on the slower 8-bit buses - so requiring the liberal use of the FWAIT instruction by the programmer.  I do vaguely remember having to add the odd FWAIT to make code that "accidentally worked" on an 8086 actually "still work" on an 8088.

Furthermore, the problems in this regard could be exacerbated if devices were not matched in each other's speed capability and the system clock, something pointed out on page 163 in the book "Upgrading and Repairing PCs".  This book also identifies the CPU/FPU chip parts that were supposed to be used in a table (see https://1drv.ms/u/s!AnT0q0DRuXBNz08wTWowKtK7l43r).  Again, this serves to underline that the data transfer between the chips was not exactly synchronous without suitable FWAIT usage, but even then could fall apart if either the CPU or the FPU was not a rated part for the same clock speed provided by the system bus, and, one assumes, for each other.  I don't know if this would affect your emulations in any way, or at all.  Do you emulate the behaviour at the system bus signal and clock-cycle level - and thus ignore or get scuppered by the idiosyncrasies of asynchronous chip interactions?

If your emulation does get down and dirty at this level, I'd recommend reading the 8087 chapters in the [iAPX 86 88 186 188 Users Manual 1985](http://bitsavers.trailing%2Dedge.com/components/intel/8086/210912%2D001_iAPX_86_88_186_188_Users_Manual_1985.pdf)  -  and doing this comparatively with the 80286 and 80386 versions of this manual.

Also, have a look at:  [Intel ASM86 Language Reference Manual Mar85](http://bitsavers.trailing%2Dedge.com/pdf/intel/ISIS_II/121703%2D003_ASM86_Language_Reference_Manual_Mar85.pdf) page 6-187.  

In fact, [this reference](https://userpages.umbc.edu/~squire/intel_book.pdf) on page 536 states that the ability to store the status word to the AX register was also provided in the 80187, not just 80287 and later.  And on page 541, this manual discusses another aspect of the 80287 and 80387 and their use of IO ports for communication between CPU and FPU - now that's something I'd never heard about!  :)  On page 547 it specifically eliminates the 8087 from supporting FSTSW AX - which I would take to imply that an invalid instruction interrupt would occur if the 8088 or 8086 encountered the opcodes for this instruction.

**Comment by BigAlUK99** [comment]
Kudos, by the way, for choosing CHKCOP as your test bed.   I shall add that to my armoury :)
https://www.pcorner.com/list/DIAGS/CHKCOP.ZIP/INFO/

But note the banner on this page:

     INTEL’s Math Coprocessor Diagnostics XT/AT/386DX & SX

...It doesn't mention the IBM 5150 PC (although to be fair the XT was an 8088/87 too).
But I don't know whether that might also be significant.  Perhaps by "XT" it actually meant "XT-286"?

**Comment by BigAlUK99** [comment]
So in summary, the instructions
```
   FSTSW mem                 ; Opcode  98 DD m7rm 
   FNSTSW mem                ; Opcode  90 DD m7rm   
```
were valid on an 8087, but the instructions
```
   FSTSW AX                  ; Opcode  9B DF E0
   FNSTSW AX                 ; Opcode  DF E0 (or indeed 90 DF E0)
```
were not.  An 80287 could execute all of these, but you won't find an 80287 on an IBM PC 5050, IBM PC 5140 Convertible,  IBM PC 5155 Portable, IBM PC 5160 XT, IBM 4860 PCjr, or the PS/2 8086-based models.  So those latter two instructions will (and should) always fail if you run CHKOP on those machines.

**Comment by waltje** [comment]
I have real machines, including one that has a CCR cassette drive (as Obat knows.)

Fred
__
Sent from my Samsung Note4 LTE smartphone


-------- Original message --------

**Comment by BigAlUK99** [comment]
Ok - I was just updating those download links.  It turns out that version 1.43 was even more buggy than I thought, as it broke something that was still working in version 1.42.   So... courtesy of my SVN Config control, please use version 1.42 - whether via floppy or cassette.  :)
https://1drv.ms/u/s!AnT0q0DRuXBNz0YDZhUQar1ggt5c - 160K Image (IMG) file
https://1drv.ms/u/s!AnT0q0DRuXBNz0eJ_JYidQdKOYVm - PCE-compatible CAS
https://1drv.ms/u/s!AnT0q0DRuXBNz1J3bF2S_5NpuZMI - Human readable .BAS text file  (you have to actually download it to read the whole file - thanks to the way MS Live works).

Thanks in advance, waltje   :)

**Comment by OBattler** [comment]
It turns out CHKCOP was executing FSTSW AX because it was misidentifying the 8087, because I didn't have the FENI and FDISI instructions (used for enabling/disabling interrupts) implemented at all. Now it works correctly in that regard. It still fails at the addition test, though, which I need to figure out why.

Meanwhile, 87TEST passes most of the tests, just fails Test 7 and the stress test, I believe.

Also, the FPU is optional in the emulator Settings, for any CPU that's not at least a 486DX (because 486DX was the first to have a built-in FPU). Also, it's interesting to note that the "487" for use with 486SX was in fact a 486DX and took over the main CPU functionality as well.

With regards to potential CPU-FPU clock mismatch, that is not a concern as the emulation is designed such that the CPU runs the FPU instructions if it sees any ESCAPE instruction and the FPU is enabled in Settings.

Also, eventually I plan to write a more accurate x87 emulation for use as 8087, 287, and 387, that would be internally 80-bit, and therefore be precision-accurate, and would also be more cycle-accurate, but it's going to be for after 86Box v2.00 is out, as it's going to take a while to write that - it took me about 2 months to replace the 808x emulation with more correct emulation based on reenigne's XTCE which is turn based on Intel's microcode and reverse-engineering of the real 8088, so I expect the new x87 to take about as much time after I start doing it.

**Comment by BigAlUK99** [comment]
This sounds excellent - shall be looking forward to it :)

Also, good to know that it appears that CHKCOP really was targeted at ordinary XT use (not just XT-286).  I shall assume that an ordinary IMB PC 5150 is therefore ok as a host, despite not being mentioned in that web-banner.

From your comment about CPU-FPU clock, I assume that this basically means that you will be enforcing synchronization - so code will work identically even if no FWAIT instructions were programmed into the app running on the emulator.  This will of course mean that any code optimizations that try to take advantage of CPU-FPU real-time parallelism will become somewhat futile,  ...but I guess one has to draw the line somewhere :)  So it all sounds good to me.  

Remember tho' that FWAIT isn't really just about "allowing time", it was also (supposed to be) about synchronizing two parallel devices so that only one will control the bus at any one time and both can be engaged in the resultant mutual transaction.  I'm assuming that the emulation will be more trivial than this, making FWAIT redundant as such.  Apparently a lot of this additional "weirdness" about FWAIT went away with the 80287, because Intel modified the behaviour of synchronization to be more forgiving (an implicit wait on the start of any FPU instruction - see the second bullet point in [this article](http://www.cpu%2Dworld.com/CPUs/80287/index.html).  But for the 8087, this will mean that anybody who writes code on the emulator may well find all kinds of stuff going wrong if they later run that code on a real machine - because they forgot to write some crucial FWAITs somewhere.  

...But who does that anyway?   :)

Remember to write that caveat on the box, and all will be well.

**Comment by BigAlUK99** [comment]
P.S.  I should probably also mention that the little article I quoted there is a tinsy-bit flawed.  It claims that the 8087 and 80287 are "object-code-compatible".  But we now know that the 8087 doesn't do FSTSW AX, don't we.   :)   So it would probably make sense to corroborate this thing about the 80287's implicit waits from another source.  Also, the article claims an 80386 should work with an 80287.  Really?  Hmm.

**Comment by OBattler** [comment]
That probably means the 287 is backwards-compatible with 8087, which it is. 387 is less so, as projective infinity is gone, I believe.

In any case, it appears that I can't fix CHKCOP's tests for time being as they rely on a level of precision the current x87 emualation can not attain. I am going to write new 8087/287/387 emulation after 86Box v2.00 is out, which should fix that.

**Comment by BigAlUK99** [comment]
Yes, you are right.  Backwards-compatible would have been a better phrasing, too.  But I'm a little suspicious still about whether an 80387 would have been fully signal- and clock-compatible with an 80286 (that other claim in the article).  But I guess if it had exactly the same pin-out...  Though not sure I'd want to rely on it.  Wanting to fit them together in a real machine sounds a little bit too desperate to me.

Precision - hmm - that's another interesting point.  Were there differences in precision between 8087, 80287, 80387, etc?  That's going to be a scary prospect from the emulation point of view isn't it?  How do you go about verifying something like that?  Not to mention that classic from Intel where they manged to widely release a flawed design where the occasional calculation would go wrong.  :) 

I also remember vaguely that there were "standard practice" routines that used to rely upon behavioural quirks in order to detect which chips were fitted, and replicating those "undocumented" quirks (documented only by the existence of such routines) will be hard enough.  

What's your view on these (potentially less critical) aspects and their impact on emulation?

**Comment by BigAlUK99** [comment]
Here's one such interesting little snippet from some Turbo Pascal code that I last updated back in 1995:
```
{-------------------------------------------------}
{ Coprocessor Tests and standardised quick-result }
{-------------------------------------------------}
{ This routine evaluates the existence of a particular coprocessor type when
  it is first called.  On subsequent calls, it refers to a stored constant
  rather than reevaluating.
}
  Const EvaluatedCoprocessor : tIdentity = Null_ID;
  Var   Scratch : word;
  Function Coprocessor : tIdentity;
  begin
     If EvaluatedCoprocessor = Null_ID  then
       asm
        { Test for the correct status word flags after finit }
          {finit}                           db $DB,$E3
           mov    [Scratch],$55AA
          {fstsw  word ptr[Scratch]}        db $DD,$3E; dw Scratch;
           cmp    byte ptr [Scratch],0
           jne    @No87
          {fstcw  word ptr[Scratch]}        db $D9,$3E; dw Scratch;
           mov    ax,[Scratch]
           and    ax,$0103
           cmp    ax,$0003
        { Test for 8087 handling of the disable interrupts flag after fdisi }
           and    [Scratch],$FF7F
          {fldcw  word ptr[Scratch]}        db $D9,$2E; dw Scratch;
          {fdisi}                           db $DB,$E1;
          {fstcw  word ptr[Scratch]}        db $D9,$3E; dw Scratch;
           db     fwait
           test   [Scratch],$0080
           jnz    @8087
        { Test for equality of positive and negative infinity on an 80287 }
          {finit}                           db $DB,$E3
          {fld1}                            db $D9,$E8
          {fldz}                            db $D9,$EE
          {fdivp  st(1),st}                 db $DE,$F9
          {fld    st(0)}                    db $D9,$C0
          {fchs}                            db $D9,$E0
          {fcompp st(1)}                    db $DE,$D9
          {fstsw  word ptr[Scratch]}        db $DD,$3E; dw Scratch;
           db     fwait
           mov    ax,[scratch]
           sahf
           je     @80287
           @80387:mov al,_80387p;   jmp @Done
           @80287:mov al,_80287;    jmp @Done
           @8087: mov al,_8087;     jmp @Done
           @No87: mov al,_No87
           @Done:
           mov    EvaluatedCoprocessor,al
           mov    @Result,al
       end
       else CoProcessor := EvaluatedCoprocessor;
  end;
```

It doesn't know how to identify an 80486DX or 80487SX, or later devices, though - I did not own any to play with back then.  :)  Looking back at this code, I'm not sure what I was doing in the region of line 23
(`cmp    ax,$0003`) which is not followed by any conditional instruction.  Hmm.  Bug? Residual experiment?  Unfinished code?  I shall only know now if I have an emulator to try it on!   ...Maybe.  :p

**Comment by OBattler** [comment]
I believe some Compaq machines supported a 386 with a 287, but not with a 387, as strange as it sounds.

Also, 8087 and later all had the same precision, and in fact CHKCOP fails the same way on all of them on the emulator. They were internally 80-bit but the emulation is internally 64-bit with some hacks in the code to make some software (mostly Pentium-era games) happy.

Interesting to hear about the existence of such undocumented "quirks", I wonder if that's what CHKCOP actually fails on.

Also, 487SX/486DX is the same as 387 really, so you need to do CPU detection instead, to see if you're on a 386 or a 486. That can be done easily by testing some bits in FLAGS and EFLAGS. I believe they did no changes to the FPU until the Pentium Pro, which adds some new instructions, such as the conditional compares.

There was also also a 287-A or what it was, that CKHCOP detects the 387 as, if it can't detect IRQ 13 on error.
Edit: It's 80287 XL, 80287 XLT, or 80C287A.
Edit #2: It looks like the 287 XL was a 387 in 287 packaging.

And it looks like there was also an 187, which was a 387 with an 8087 interface, designed for use with the 186 (but not the 188 which was 8-bit). I wonder if it could be used with an 8086 (but not 8088, which is, again 8-bit)?

**Comment by waltje** [comment]
That was because Intel did not have working 387s yet, and Compaq had to move.  So, they went with 287 instead.

Fred
__
Sent from my Samsung Note4 LTE smartphone


-------- Original message --------

**Comment by OBattler** [comment]
![20190629_095205](https://user-images.githubusercontent.com/10365853/60381341-ae5ffc80-9a53-11e9-8e62-7c347b99dcc3.png)

I added the 5161 and as the screenshot shows, it works. The only thing missing of the 5150 now is the cassette. I'm going to attempt to port cassette emulation from PCE so that I can use the .CAS files linked to in this thread.

Also, I made the 1981 IBM PC use Cassette BASIC 1.00 now as well.

**Comment by BigAlUK99** [comment]
I'm chuffed that my BASIC program is proving useful.  :)
You are the first to document it being used for its intended purpose - a handy quick test harness (i.e. sanity check) for 5150 Emulator developers.

**Comment by BigAlUK99** [comment]
P.S. I'm considering whether to provide an update to the tests concerning the LPT on the graphics cards.

If the LPT is disabled (by removing a chip, as described by [John Elliot](https://www.seasip.info/VintagePC/hercplus.html#printer)) it becomes possible for another 8-bit board to exist with an LPT port that  occupies the same addresses as the one that would have normally resided on the graphics card.  Quite why someone would set out to do this isn't clear, but it does beg the question of slot counts if the only LPT ports are _not_ at the addresses expected for the graphics card's LPT device.

Where I'm going with this is to ask your opinion whether it would make sense to validate that a graphics card's LPT port actually resides on the correct address before simply assuming one can deduct a slot count?  Or would this just lessen the tool's use as a sanity check by making it far too strict?

**Comment by OBattler** [comment]
There's no way to determine what is providing the LPT port on a specific address.

**Comment by OBattler** [comment]
@BigAlUK99 I have experimentally hooked up PCE's cassette emulation to 86Box, and after some tinkering (I had to put the cassette on a timer running at 1/8 the ~1.19 MHz PIT speed), I was able to successfully load and run your .CAS file. It works quite nicely.

**Comment by BigAlUK99** [comment]
I guess I really am going to have to update it now.  :)

I'll get there eventually.  

Got rather busy of late (a) having a pacemaker fitted and (b) tinkering around with my thirty-odd emulators/VMs:  trying to standardise what's installed in the Windows ones (1.0 through to 10, missing very little), ...though sometimes I wonder why I'm doing all this.  Must be 'coz I'm fully retired now ...heh, heh.  I wonder who'll inherit all this?

Oh dear - these thoughts are all way too existential...   now, what was it I was doing?

**Comment by BigAlUK99** [comment]
Been out of action for a while (suffered heart-failure).

So I've gone from semi-retired (with income) to fully retired (but no income yet)  !!!

Have been updating all my emulators this week.  So what is the status regarding the CAS interface these days?

**Comment by OBattler** [comment]
I have it ported from PCE, but I haven't yet committed it or hooked it up to the GUI, will probably do so after my current work (mem.c/h overhaul, AT / PS/2 keyboard + controller rewrite, PIC rewrite, adding SiS 40x, 460, 461, and 50x chipsets + correpsonding mchines, and going through all the other chipsets) is done.

**Comment by BigAlUK99** [comment]
Really looking forward to it: the most recent stable build is already looking sweet. you've clearly been really busy and productive in the past year!  :)

**Comment by Torinde** [comment]
Related:
- #4548 
- #5248
